okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt

```diff
@@ -37,6 +38,10 @@ open class Jdk9Platform : Platform() {
 
     sslParameters.applicationProtocols = names.toTypedArray()
 
+    if (hostname != null) {
+      sslParameters.serverNames = listOf(SNIHostName(hostname))
```

Comment:
My read is we haven’t done SNI correctly on non-Android platforms. . . . forever? Wow!

Reply:
It was deliberate. Its meant to be inbuilt

Reply:
Maybe our usage of the Socket API breaks this?

```
        // Create the wrapper over the connected socket.
        val sslSocket = route.address.sslSocketFactory.createSocket(
          rawSocket,
          route.address.url.host,
          route.address.url.port,
          true /* autoClose */
        ) as SSLSocket
```        
        
Surely the host here should signal SNI, which was enable for clients by default in JDK 7?

Reply:
I think it doesn’t, and I’m horrified. In particular I think we needed to manually opt-in with the `SNIHostName` thing.

Reply:
OK, so I don't think it's this bad.  I think we don't understand the full set of things going on.  Your test ends up with "127.0.0.1" 

<img width="692" alt="image" src="https://user-images.githubusercontent.com/231923/210287713-f0cc211a-01bd-4c1e-9038-2a1cb781a649.png">

While with an external site it's the correct hostname

<img width="799" alt="image" src="https://user-images.githubusercontent.com/231923/210287770-8e461ae1-272f-40ec-9941-fa38a11433fd.png">

Reply:
localhost is not a valid SNI name.

<img width="912" alt="image" src="https://user-images.githubusercontent.com/231923/210287926-62e09014-7fb0-49bf-b5be-f2a08aaf1752.png">

Reply:
Spec: https://datatracker.ietf.org/doc/html/rfc6066#page-6

I assume it's this

>  "HostName" contains the fully qualified DNS hostname of the server, as understood by the client.